File: compiler/bench/median3.c
// expected: 50000005000000
// Median of 3 benchmark - 10M iterations

#include <stdio.h>

#define ITERS 10000000

int median3(int a, int b, int c) {
    if (a > b) { int t = a; a = b; b = t; }
    if (b > c) { int t = b; b = c; c = t; }
    if (a > b) { int t = a; a = b; b = t; }
    return b;
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        sum += median3(i, i + 1, i + 2);
    }
    return sum;
}

int main(void) {
    printf("%ld\n", bench());
    return 0;
}

File: compiler/bench/median3.fs
\ expected: 50000005000000
\ Median of 3 benchmark - 10M iterations

10000000 constant ITERS

: median3 ( a b c -- median )
  rot 2dup < if swap then    \ ensure a >= b on stack: c a' b' where a'>=b'
  rot 2dup < if swap then    \ ensure c >= b: a' c' b' where c'>=b'
  drop max ;                  \ max of the two smaller ones

: bench ( -- sum )
  0
  ITERS 0 do
    i i 1+ i 2 + median3 +
  loop ;

: main
  bench . cr ;
